doc/crypt.tex

```diff
@@ -5579,12 +5595,71 @@ \subsection{Signature Verification}
 This function validate the ECDSA signature as \textit{ecc\_verify\_hash} only the signature input format
 follows \url{https://tools.ietf.org/html/rfc7518#section-3.4}.
 
+\index{ecc\_verify\_hash_ex()}
+\begin{verbatim}
+int ecc_verify_hash_ex(const unsigned char *sig,
+                             unsigned long  siglen,
+                       const unsigned char *hash,
+                             unsigned long  hashlen,
+                        ecc_signature_type  sigformat,
+                                       int *stat,
+                                   ecc_key *key);
+\end{verbatim}
+
+This function validates an ECDSA signature as \textit{ecc\_verify\_hash} but with a choice of signature formats.
+
 {\bf BEWARE:} With ECC if you try to sign a hash that is bigger than your ECC key you can run into problems. The math
 will still work, and in effect the signature will still work.  With ECC keys the strength of the signature is limited
 by the size of the hash, or the size of the key, whichever is smaller.  For example, if you sign with SHA256 and a
 P--192 key, you have in effect 96--bits of security. The library will not warn you if you make this mistake, so it
 is important to check yourself before using the signatures.
 
+\subsection{Public Key Recovery}
+\index{ecc\_recover\_key()}
+\begin{verbatim}
+int ecc_recover_key(const unsigned char *sig,
+                          unsigned long  siglen,
+                    const unsigned char *hash,
+                          unsigned long  hashlen,
+                                    int  recid,
+                     ecc_signature_type  sigformat,
+                                ecc_key *key);
+\end{verbatim}
+
+This function will recover (a) public key from the ECDSA signature in the array pointed to by \textit{sig} of length \textit{siglen} octets, the message digest
+pointed to by the array \textit{hash} of length \textit{hashlen}, and the recovery id \textit{recid}. It will store the recovered
+key into \textit{key} and return CRYPT\_OK if recovery succeeds, or an error if recovery fails.
+This is for compatibility with the (v,r,s) signatures used in Ethereum, where public keys are not explicitly shared,
+only the parity of the public key. For curves like secp256k1, recid will take values of 0 or 1, corresponding to the
+parity of the public key's y coordinate. For curves like secp112r2, with a cofactor of 4, values 0..7 are possible,
```

Comment:
The implementation will recover keys for any of the ECC curves - this is exercised in the tests (around line 535 of `ecc_test.c`, in a loop through curve types):
```
         for (j = 0; j < 2*(1+(int)privkey.dp.cofactor); j++) {
            stat = ecc_recover_key(buf, len, data16, 16, j, LTC_ECCSIG_ANSIX962, &reckey);
            if (stat != CRYPT_OK) continue; /* last two will almost always fail, only possible if x<(prime mod order) */
            stat = _ecc_key_cmp(PK_PUBLIC, &pubkey, &reckey);
            if (stat == CRYPT_OK) found++;
         }
         if (found != 1) return CRYPT_FAIL_TESTVECTOR; /* unique match */
```

The `_ecc_test_recovery` function actually goes through all of the different curves and tests recovery - generating a `LTC_ECCSIG_RFC7518` signature and separate recovery ID (line 1084), then recovering the public key (line 1093) and checking it matches (line 1094).


The restriction on secp256k1 is only for generating `LTC_ECCSIG_ETH27` signatures, because that's the only curve used by Ethereum/Bitcoin/etc and because adding 27 to the recovery value is specific to Ethereum and gets rather confusing for curves with a cofactor other than 1, so the signature format is undefined for any other curve type.

Reply:
Oh... there _is_ an `#ifdef LTC_ECC_SECP256K` block in `_ecc_test_recovery`, because the `LTC_ECCSIG_ETH27` signature example is for secp256k1, so there's no way to run that part of the test if that curve isn't included in the build.

Reply:
As mentioned elsewhere I am for removing the restriction on secp256k1 and declare `LTC_ECCSIG_ETH27` as ethereum-like style of storing signatures as: zero-padded R value + zero-padded S value + 1 byte with recovery_id + 27.